core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java

```diff
@@ -26,38 +28,41 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-import javax.annotation.Nullable;
-
 import com.google.common.collect.ImmutableSet;
 
 /**
- * A bindable path. It holds three things:
+ * A path param extractor. It holds three things:
  * <ul>
- *   <li> A regex-compiled form of the path. It is used for matching and binding.
+ *   <li> A regex-compiled form of the path. It is used for matching and extracting.
  *   <li> A skeleton of the path. It is used for duplication detecting.
  *   <li> A set of variables declared in the path.
  * </ul>
  *
- * @see DynamicHttpService
- * @see Path
  */
-final class DynamicPath {
+public final class PathParamExtractor {
 
-    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/]+)+");
+    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/{}:]+|/:[^/{}]+|/\\{[^/{}]+})+/?");
 
     /**
-     * Create a {@link DynamicPath} instance from given {@code path}.
+     * Create a {@link PathParamExtractor} instance from given {@code path}.
+     * @param path the {@link String} that contains path params.
+     *             e.g. {@code /users/{name}} or {@code /users/:name}
+     *
+     * @throws IllegalArgumentException if the {@code path} is invalid.
      */
-    static DynamicPath of(String path) {
-        if (!VALID_PATTERN.matcher(path).matches()) {
+    public static PathParamExtractor of(String path) {
+        requireNonNull(path, "path");
+
+        final String matchPath = path.charAt(0) == '/' ? path : "/" + path;
+        if (!VALID_PATTERN.matcher(matchPath).matches()) {
             throw new IllegalArgumentException(
-                    String.format("%s: Invalid DynamicPath (valid: %s)", path, VALID_PATTERN));
+                    String.format("%s: Invalid PathParamExtractor (valid: %s)", path, VALID_PATTERN));
```

Comment:
`throw new IllegalArgumentException("pathPattern: " + pathPattern);` should be enough.